compiler/rustc_mir_transform/src/lib.rs

```diff
@@ -394,6 +395,27 @@ fn inner_mir_for_ctfe(tcx: TyCtxt<'_>, def: LocalDefId) -> Body<'_> {
     body
 }
 
+fn const_prop_lint(tcx: TyCtxt<'_>, def: LocalDefId) {
+    let (body, _) = tcx.mir_promoted(def);
+    let body = body.borrow();
+
+    let mir_borrowck = tcx.mir_borrowck(def);
+
+    // If there are impossible bounds on the body being const prop linted,
+    // the const eval logic used in const prop may ICE unexpectedly.
+    let predicates = tcx
+        .predicates_of(body.source.def_id())
+        .predicates
+        .iter()
+        .filter_map(|(p, _)| if p.is_global() { Some(*p) } else { None });
+    if !traits::impossible_predicates(tcx, traits::elaborate(tcx, predicates).collect())
```

Comment:
This duplication is somewhat unfortunate. We could make `const_prop_lint` return whether there were impossible predicates and then bail out.

Or we perform this in borrowck and store the information in borrowck results.

If perf doesn't find this problematic, we could just deduplicate by moving the logic into a function

Reply:
It may be interesting to look at error taint before trying to check for impossible predicates, as that's faster. 
IIUC, this impossible predicates check is more of a heuristic to avoid diagnostic spam, as it does not try to prove satisfiability, does it?

Reply:
> It may be interesting to look at error taint before trying to check for impossible predicates, as that's faster.

there are no errors usually. With `trivial_bounds` you may have a broken function that will only error if called, never out of another reason.